src/hotspot/share/gc/g1/g1ConcurrentMark.cpp

```diff
@@ -586,24 +586,49 @@ class G1ClearBitMapTask : public AbstractGangTask {
   private:
     G1CMBitMap* _bitmap;
     G1ConcurrentMark* _cm;
+
+  bool has_aborted() {
+    if (_cm != NULL) {
+      _cm->do_yield_check();
+      return _cm->has_aborted();
+    }
+    return false;
+  }
+
+  HeapWord* region_clear_limit(HeapRegion* r) {
+    // During a Concurrent Undo Mark cycle, the _next_mark_bitmap is  cleared
+    // without swapping with the _prev_mark_bitmap. Therefore, the per region
+    // next_top_at_mark_start and live_words data are current wrt
+    // _next_mark_bitmap. We use this information to only clear ranges of the
+    // bitmap that require clearing.
+    if (_cm != NULL && _cm->cm_thread()->in_undo_mark()) {
+      // No need to clear bitmaps for empty regions.
+      if (_cm->live_words(r->hrm_index()) == 0) {
+        assert(_bitmap->get_next_marked_addr(r->bottom(), r->end()) == r->end(), "Should not have marked bits");
+        return r->bottom();
+      }
+      // Otherwise, we only clear up to the next_top_at_mark_start
+      assert(_bitmap->get_next_marked_addr(r->next_top_at_mark_start(), r->end()) == r->end(), "Should not have marked bits above ntams");
+      return r->next_top_at_mark_start();
+    }
+    return r->end();
+  }
+
   public:
     G1ClearBitmapHRClosure(G1CMBitMap* bitmap, G1ConcurrentMark* cm) : HeapRegionClosure(), _bitmap(bitmap), _cm(cm) {
```

Comment:
Since this code is so highly specific to clearing `_next_bitmap` (and depending on current state which TAMS is the correct one for a given region) I would prefer if the code did not have configurable `bitmap` too.